### ISSUE TITLE
"variable does not need to be mutable" not detected

### ISSUE THREAD

**Issue by BO41** [rust-lang/rust#61424]
(sorry if this is an duplicate or the wrong repo :innocent: )

```rust
fn main() {
    let mut name;
    name = format!("foo{}", "bar");
    dbg!(name);
}
```
([Playground](https://play.integer32.com/?version=stable&mode=debug&edition=2018&gist=50169a334a1863bd3aaed60d83222ba6))

Current behavior:
no warning

Expected behavior:
```
warning: variable does not need to be mutable
 --> src/main.rs:2:9
  |
2 |      let mut name;
  |          ----^
  |          |
  |          help: remove this `mut`
  |
  = note: #[warn(unused_mut)] on by default
```

The variable clearly doesn't need to be mutable, still, nothing is detected and no warning is emitted.

Rust 1.35.0
(Using the [playground](https://play.integer32.com/?version=nightly&mode=debug&edition=2018&gist=0d88ecdb9fa002a3b81c7e7fc48572ab), also in `1.37.0-nightly` (`2019-05-31 7840a0b753a065a41999`))

**Comment by czipperz** [comment]
The warning appears again if you switch to 2015 edition.  Huh.  Nice catch!

**Comment by czipperz** [comment]
I found that Stable 2015 edition produces the warning and everything else doesn't.

```
| Warning | Optimization | Channel | Edition |
|---------+--------------+---------+---------|
| Present | Debug        | Stable  |    2015 |
| Bug!    | Debug        | Beta    |    2015 |
| Bug!    | Debug        | Nightly |    2015 |
| Present | Release      | Stable  |    2015 |
| Bug!    | Release      | Beta    |    2015 |
| Bug!    | Release      | Nightly |    2015 |
| Bug!    | Debug        | Stable  |    2018 |
| Bug!    | Debug        | Beta    |    2018 |
| Bug!    | Debug        | Nightly |    2018 |
| Bug!    | Release      | Stable  |    2018 |
| Bug!    | Release      | Beta    |    2018 |
| Bug!    | Release      | Nightly |    2018 |
```

**Comment by jonas-schievink** [comment]
cc @matthewjasper (likely caused by NLL migrate mode, which was enabled in https://github.com/rust-lang/rust/pull/59114 for the 2015 edition)

**Comment by matthewjasper** [comment]
Yes, looks like another nll bug. If someone wants to have a go at this, the issue is that we are treating DropAndReplace differently to Assign for the purposes of this lint.

**Comment by czipperz** [comment]
I can try doing it I just don't know here to begin.

**Comment by jonas-schievink** [comment]
@czipperz I don't know this code too well, but take a look at `src/librustc_mir/borrow_check/used_muts.rs`. The first thing I'd try is adding a case for `DropAndReplace` to the match in `visit_statement` that mirrors what's done for `Assign`.

**Comment by czipperz** [comment]
Makes sense thank you.  I was looking at librustc_lint and couldn't find anything.
